Create Profile.js

MilesDev91 opened the pull request:
Model for the user profile. Tried to include everything needed to start the app with all features, but is easily expandable. For example, favorite pets is a potential feature, but not necessary at the beginning.

### What this PR does (required):
- Models the user's profile

### Any information needed to test this feature (required):
- For now, it is just the model

I am curious as to how I did with the model. I have not worked with models in a long time, and I am not sure if I put too much in or not enough.

MilesDev91 committed:
Create Profile.js

Model for the user profile. Tried to include everything needed to start the app with all features, but is easily expandable. For example, favorite pets is a potential feature, but not necessary at the beginning.

ugur-akin commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+    type: String,
+    required: true
+  },
+  // To allow similar names
+  id: {
+    type: Number,
+    required: true
+  },
```
Good thought, but mongoose by default adds a unique `_id` field to schemas which can be used to distinguish different profiles 👍🏼 ([docs](https://mongoosejs.com/docs/guide.html#_id))

ugur-akin commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+    type: Number,
+    required: true
+  },
+  availability: {
+    unavailableDates: [Date],
+    extraAvailability: [Date],
+    weeklyAvailability: {
```
This is pretty nice 💯

ugur-akin commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+    unavailableDates: [Date],
+    extraAvailability: [Date],
+    weeklyAvailability: {
+      // These will be strings of date time ranges
+      monday: [String],
+      tuesday: [String],
+      wednesday: [String],
+      thursday: [String],
+      friday: [String],
+      saturday: [String],
+      sunday: [String]
```
Alternatively, you could use a subdocument to set ranges here. ([docs](https://mongoosejs.com/docs/subdocs.html#subdocuments))

ugur-akin commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+    type: Number,
+    required: true
+  },
+  availability: {
+    unavailableDates: [Date],
+    extraAvailability: [Date],
+    weeklyAvailability: {
+      // These will be strings of date time ranges
+      monday: [String],
+      tuesday: [String],
+      wednesday: [String],
+      thursday: [String],
+      friday: [String],
+      saturday: [String],
+      sunday: [String]
+    }
```
This is a pretty good and fairly complex availability model, I like it ❤️ In order to avoid confusion though, you could add more small descriptions/use cases like the one on `line 12` 👍🏼

ugur-akin commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+    hourlyRate: Number,
+    Rating: Number
+  },
+  requests: [{
+    // Decides whether the request is for this user's services or sent from this user to request services
+    isSitter: { type: Boolean, required: true },
+    startDate: Date,
+    endDate: Date
+  }],
+  booking: {
+    appointments: [{
+      date: { type: Date, required: true },
+      owner: { type: String, required: true },
+      pet: {
+        breed: { type: String, required: true },
+        name: { type: String, required: true }
+      }
+    }]
+  },
```
It's better for these two to have their own models, they don't really belong in the profile model 👍🏼

ugur-akin commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+      }
+    }]
+  },
+  favoriteUsers: [String],
```
This should be a reference to a model, I think either `Profile` or `User` would work depending your approach. ([mongoose docs](https://mongoosejs.com/docs/populate.html) about reference/population)

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+    required: true
+  },
+  // To allow similar names
+  id: {
```
I don't think we need this field as Mongoose will automatically generate an `id` for us each time a new document is created

ugur-akin commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+  }]
+})
+
+module.exports = Profile = mongoose.model("profile", profileSchema)
```
```suggestion
module.exports = Profile = mongoose.model("Profile", profileSchema)
```
Functionally doesn't make a difference but it's best to pass the exact name of your model to mongoose. ([docs](https://mongoosejs.com/docs/models.html#compiling))

rajivtitus replied:
```suggestion
module.exports = Profile = mongoose.model("Profile", profileSchema)
```
Nit: Please resort to using PascalCase for model names

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+    }
+  },
+  description: String,
+  sitter: {
```
Good job thinking ahead 💯

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+  },
+  description: String,
+  sitter: {
+    isSitting: { type: boolean, required: true },
```
```suggestion
    isSitting: { type: Boolean, required: true, default: false },
```
We can also add a default value to this field

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+  sitter: {
+    isSitting: { type: boolean, required: true },
+    hourlyRate: Number,
+    Rating: Number
```
We will have a separate model for Rating/Reviews, you can remove it here :)

ugur-akin replied:
`Rating`s will also have their own model, but I'm not against having a field for "Overall rating" for a profile. That said, `Rating` should be in `camelCase` for consistency.

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+    hourlyRate: Number,
+    Rating: Number
+  },
+  requests: [{
```
Similar to the case above, we are going to have a separate model for requests as well, you can safely remove this field too!

rajivtitus replied:
We won't require this field as the Request model will help us keep track of the owner and the sitter within a booking request.

ugur-akin reviewed (commented):
Good job with this 🚀 🎉 There are a few minor concerns but the model looks solid overall - really like the availability structure 🙌🏼 

As a bonus, it would be nice to see some validations as well (e.g. `end > start` for date ranges or max length for descriptions...) 😎

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,60 @@
+    startDate: Date,
+    endDate: Date
+  }],
+  booking: {
```
Similar to the case above, we are going to have a separate model for Requests as well, you can safely remove this field too!

rajivtitus commented:
Great work thinking ahead! 💯 
Just some comments from Ugur and myself, please review and address them!

MilesDev91 committed:
Update Profile.js

Added validation for several fields.

Clarified rating field.

Added separate schema for time ranges and its own validation.

Referenced profile instead of a string for favorite users.

MilesDev91 commented:
Hopefully things are looking better now.

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,59 @@
+const mongoose = require("mongoose")
+
+const timeRangeSchema = new mongoose.Schema({
```
👍🏽

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,59 @@
+    max: 120
+  },
+  sitter: {
+    isSitting: { type: boolean, required: true, default: false },
```
```suggestion
    isSitting: { type: Boolean, required: true, default: false },
```

rajivtitus reviewed (approved):
Great work with those changes! Just a small comment to address before merging, thanks! 🚀

MilesDev91 committed:
Update Profile.js
